pkg/controller/changetierrequest/changetierrequest_controller.go

```diff
@@ -74,5 +85,132 @@ func (r *ReconcileChangeTierRequest) Reconcile(request reconcile.Request) (recon
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: requeueAfter,
+		}, nil
+	}
+
+	err = r.changeTier(reqLogger, changeTierRequest, request.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Change of the tier is completed")
+	return reconcile.Result{
+		Requeue:      true,
+		RequeueAfter: r.config.GetDurationBeforeChangeRequestDeletion(),
+	}, r.setStatusChangeComplete(changeTierRequest)
+}
+
+func (r *ReconcileChangeTierRequest) checkTransitionTimeAndDelete(log logr.Logger, changeTierRequest *toolchainv1alpha1.ChangeTierRequest, completeCond toolchainv1alpha1.Condition) (bool, time.Duration, error) {
+	log.Info("the ChangeTierRequest is completed so we can ignore it")
+	timeFromLastChange := time.Since(completeCond.LastTransitionTime.Time)
+
+	if timeFromLastChange >= r.config.GetDurationBeforeChangeRequestDeletion() {
+		log.Info("the ChangeTierRequest is completed for longer time than '%s' so it's ready to be deleted", r.config.GetDurationBeforeChangeRequestDeletion().String())
+		if err := r.client.Delete(context.TODO(), changeTierRequest, &client.DeleteOptions{}); err != nil {
+			return false, 0, errs.Wrapf(err, "unable to delete ChangeTierRequest object '%s'", changeTierRequest.Name)
+		}
+		return true, 0, nil
+	}
+	return false, r.config.GetDurationBeforeChangeRequestDeletion() - timeFromLastChange, nil
+}
+
+func (r *ReconcileChangeTierRequest) changeTier(log logr.Logger, changeTierRequest *toolchainv1alpha1.ChangeTierRequest, namespace string) error {
+	mur := &toolchainv1alpha1.MasterUserRecord{}
+	murName := types.NamespacedName{Namespace: namespace, Name: changeTierRequest.Spec.MurName}
+	if err := r.client.Get(context.TODO(), murName, mur); err != nil {
+		return r.wrapErrorWithStatusUpdate(log, changeTierRequest, r.setStatusChangeFailed, err, "unable to get MasterUserRecord with name %s", changeTierRequest.Spec.MurName)
+	}
+
+	nsTemplateTier := &toolchainv1alpha1.NSTemplateTier{}
+	tierName := types.NamespacedName{Namespace: namespace, Name: changeTierRequest.Spec.TierName}
+	if err := r.client.Get(context.TODO(), tierName, nsTemplateTier); err != nil {
+		return r.wrapErrorWithStatusUpdate(log, changeTierRequest, r.setStatusChangeFailed, err, "unable to get NSTemplateTier with name %s", changeTierRequest.Spec.TierName)
+	}
+
+	namespaces := make([]toolchainv1alpha1.NSTemplateSetNamespace, len(nsTemplateTier.Spec.Namespaces))
+	for i, ns := range nsTemplateTier.Spec.Namespaces {
+		namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
+			Type:     ns.Type,
+			Revision: ns.Revision,
+		}
+	}
+	newNsTemplateSet := toolchainv1alpha1.NSTemplateSetSpec{
+		TierName:   changeTierRequest.Spec.TierName,
+		Namespaces: namespaces,
+	}
+	changed := false
+	for i, ua := range mur.Spec.UserAccounts {
+		if changeTierRequest.Spec.TargetCluster != "" {
+			if ua.TargetCluster == changeTierRequest.Spec.TargetCluster {
```

Comment:
these 2 `if` statements could be combined

Reply:
not really. I need to cover the `else` part appropriately. There are these cases possible
1. targetCluster is specified && is equal to the one in userAccount spec - then:
   * we need to change the tier inside of the userAccount and exit the loop 
2. targetCluster is specified && is NOT equal to the one in userAccount spec
   * we need to skip this UserAccount and wait until the targetCluster is matched
3. targetCluster is NOT specified
   * we need to change tier in all UserAccounts

If I combined it, then I wouldn't be able to cover all the cases

Reply:
ok, got it.